rpcs3/Emu/Cell/PPUTranslator.cpp

```diff
@@ -4004,11 +4004,9 @@ void PPUTranslator::FRSP(ppu_opcode_t op)
 void PPUTranslator::FCTIW(ppu_opcode_t op)
 {
 	const auto b = GetFpr(op.frb);
-	//const auto sat_l = m_ir->CreateFCmpULT(b, ConstantFP::get(GetType<f64>(), -std::pow(2, 31))); // TODO ???
-	//const auto sat_h = m_ir->CreateFCmpOGE(b, ConstantFP::get(GetType<f64>(), std::pow(2, 31)));
-	//const auto converted = m_ir->CreateFPToSI(FP_SAT_OP(sat_l, b), GetType<s64>());
-	//SetFpr(op.frd, m_ir->CreateSelect(sat_h, m_ir->getInt64(0x7fffffff), converted));
-	SetFpr(op.frd, Call(GetType<s32>(), "llvm.x86.sse2.cvtsd2si", m_ir->CreateInsertElement(GetUndef<f64[2]>(), b, uint64_t{0})));
+	const auto res = Call(GetType<s32>(), "llvm.x86.sse2.cvtsd2si", m_ir->CreateInsertElement(GetUndef<f64[2]>(), b, u64{0}));
+	const auto clamp_needed = m_ir->CreateAnd(m_ir->CreateICmpEQ(res, m_ir->getInt32(INT32_MIN)), m_ir->CreateFCmpOGE(b, ConstantFP::get(GetType<f64>(), std::exp2l(31))));
```

Comment:
Unfortunately, this is very inefficient way of checking for overflow, because it depends on cvt instruction result.

Reply:
This is intentional for FCTIFW/D, it avoids rounding errors by doing `select<s32>(rb >= INT32_MAX, INT32_MIN, cast<s32>(rb));
because ranges between f64{INT32_MAX} > rb >= f64{INT32_MAX} - 0.5 which should be rounded up to INT32_MAX by cvt instruction are missed.

I can avoid it with FCTIFW/DZ though.

Reply:
Fixed.